Assets/__Scripts/Beatmap/Animations/PointDefinition.cs

```diff
@@ -0,0 +1,285 @@
+                {
+                    prev = m;
+                }
+                else
+                {
+                    next = m;
+                }
+            }
+        }
+
+        public int CompareTo(PointDefinition<T> other)
+        {
+            // TODO: might be able to cheese previous/next with this
+            return StartTime.CompareTo(other.StartTime);
+        }
+
+        public class PointData : IComparable<PointData>
+        {
+            public PointData(Parser parser, JSONArray data, float tbegin = 0, float tend = 0)
+            {
+                Value = parser(data, out int i);
+                var len = data.Count;
+                if (len > i)
+                {
+                    // Track or Path animation
+                    Time = (tend == 0)
+                        ? data[i++].AsFloat
+                        : Mathf.LerpUnclamped(tbegin, tend, data[i++]);
+                }
+                else
+                {
+                    // WTF Jevk
+                    Time = 0;
+                }
+                Easing = global::Easing.Linear;
+                Lerp = PointDataInterpolators.LinearLerp<T>;
+
+                for (; i < len; ++i)
+                {
+                    string str = data[i];
+                    if (str[0] == 'e')
+                    {
+                        Easing = global::Easing.Named(str);
+                    }
+                    if (str == "splineCatmullRom")
+                    {
+                        Lerp = PointDataInterpolators.CatmullRomLerp<T>;
+                    }
+                    if (str == "lerpHSV")
+                    {
+                        Lerp = PointDataInterpolators.HSVLerp<T>;
+                    }
+                }
+            }
+
+            public T Value { get; }
+            public float Time { get; }
+            public Func<float, float> Easing { get; }
+            public InterpolationHandler Lerp { get; }
+
+            public int CompareTo(PointData other)
+            {
+                return Time.CompareTo(other.Time);
+            }
+        }
+    }
+
+    public class PointDataInterpolators
+    {
+        public static T LinearLerp<T>(PointDefinition<T>.PointData[] points, int prev, int next, float time) where T : struct
+        {
+            return LinearLerpFunc<T>()(points[prev].Value, points[next].Value, time);
+        }
+
+        public delegate T LerpFunc<T>(T prev, T next, float time);
+        static LerpFunc<float> LinearFloat = new LerpFunc<float>(Mathf.LerpUnclamped);
+        static LerpFunc<Color> LinearColor = new LerpFunc<Color>(Color.LerpUnclamped);
+        static LerpFunc<Vector3> LinearVector3 = new LerpFunc<Vector3>(Vector3.LerpUnclamped);
+        static LerpFunc<Quaternion> LinearQuaternion = new LerpFunc<Quaternion>(Quaternion.SlerpUnclamped);
+
+        public static LerpFunc<T> LinearLerpFunc<T>() where T : struct
+        {
+            // I hate C#
```

Comment:
I believe newer versions of C# allow you to switch like this

```cs
return T switch
{
    float n => (LerpFunc<T>)(object)(LinearFloat),
    // ...
}
```

But Unity's Mono runtime doesnt support that language feature 🙃